backend/src/apiserver/storage/experiment_store.go

```diff
@@ -26,33 +27,70 @@ type ExperimentStore struct {
 	resourceReferenceStore *ResourceReferenceStore
 }
 
-func (s *ExperimentStore) ListExperiments(opts *list.Options) ([]*model.Experiment, string, error) {
-	errorF := func(err error) ([]*model.Experiment, string, error) {
-		return nil, "", util.NewInternalServerError(err, "Failed to list experiments: %v", err)
+// Runs two SQL queries in a transaction to return a list of matching experiments, as well as their
+// total_size. The total_size does not reflect the page size.
+func (s *ExperimentStore) ListExperiments(opts *list.Options) ([]*model.Experiment, int, string, error) {
+	errorF := func(err error) ([]*model.Experiment, int, string, error) {
+		return nil, 0, "", util.NewInternalServerError(err, "Failed to list experiments: %v", err)
 	}
 
-	sql, args, err := opts.AddToSelect(sq.Select("*").From("experiments")).ToSql()
+	// SQL for getting the filtered and paginated rows
+	sqlBuilder := opts.AddFilterToSelect(sq.Select("*").From("experiments"))
+	rowsSql, rowsArgs, err := opts.AddPaginationToSelect(sqlBuilder).ToSql()
 	if err != nil {
 		return errorF(err)
 	}
 
-	rows, err := s.db.Query(sql, args...)
+	// SQL for getting total size. This matches the query to get all the rows above, in order
+	// to do the same filter, but counts instead of scanning the rows.
+	sizeSql, sizeArgs, err := opts.AddFilterToSelect(sq.Select("count(*)").From("experiments")).ToSql()
 	if err != nil {
 		return errorF(err)
 	}
-	defer rows.Close()
 
+	// Use a transaction to make sure we're returning the total_size of the same rows queried
+	tx, err := s.db.Begin()
+	if err != nil {
+		glog.Errorf("Failed to start transaction to list jobs")
+		return errorF(err)
+	}
+
+	rows, err := tx.Query(rowsSql, rowsArgs...)
+	if err != nil {
+		tx.Rollback()
+		return errorF(err)
+	}
 	exps, err := s.scanRows(rows)
 	if err != nil {
+		tx.Rollback()
+		return errorF(err)
+	}
+	rows.Close()
```

Comment:
can we follow the style here and use `defer close()` next to where rows is assigned, after error is handled?
http://go-database-sql.org/retrieving.html

Reply:
same as other places in this PR

Reply:
Actually I remember now why I had to do this. In order for multiple queries to work within a transaction, the first query's rows have to be closed before the second one's are opened, otherwise the query just fails.
I didn't find any docs on this online, other than [PostgreSQL comment](https://github.com/lib/pq/issues/81#issuecomment-229598201) that mentioned the same thing. Let me know if you know a way around it.